libraries/libfc/test/crypto/test_bls.cpp

```diff
@@ -0,0 +1,219 @@
+    BOOST_CHECK_EQUAL(p.equal(p_res), true);
+} FC_LOG_AND_RETHROW();
+
+BOOST_AUTO_TEST_CASE(g1_exp_garbage) try {
+    fp x({0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF});
+    g1 p({x, x, x});
+    BOOST_CHECK_EQUAL(x.isValid(), false);
+    BOOST_CHECK_EQUAL(p.isOnCurve(), false);
+    BOOST_CHECK_EQUAL(p.inCorrectSubgroup(), false);
+    array<uint64_t, 4> s = {0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF};
+    p = g1::multiExp({p}, {s}).value();
+    BOOST_CHECK_EQUAL(p.isOnCurve(), false);
+    BOOST_CHECK_EQUAL(p.inCorrectSubgroup(), false);
+    g1 p_res = g1::fromJacobianBytesBE(hexToBytes<144>("0x181b676153b877407d2622e91af6057f5ff445f160c178517828841670debdd61957f8d5376ddeeb1ba0a204eb1eafb007f9d1417540591155acddd91f1fb9c97da24d6eecae002c50a779372dfc247efb1823e27abbdae09fb515f390e982311239b452c1ef85156c979f981ac69208f6fd0014fa9dd66a1999df7fa4a0a4234a4cc14ec62291fd3f924b8353b326b9"), false, true).value();
+    BOOST_CHECK_EQUAL(p.equal(p_res), true);
+} FC_LOG_AND_RETHROW();
+
+BOOST_AUTO_TEST_CASE(g2_exp_garbage) try {
+    fp x({0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF});
+    g2 p({fp2({x, x}), fp2({x, x}), fp2({x, x})});
+    BOOST_CHECK_EQUAL(x.isValid(), false);
+    BOOST_CHECK_EQUAL(p.isOnCurve(), false);
+    BOOST_CHECK_EQUAL(p.inCorrectSubgroup(), false);
+    array<uint64_t, 4> s = {0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF};
+    p = g2::multiExp({p}, {s}).value();
+    BOOST_CHECK_EQUAL(p.isOnCurve(), false);
+    BOOST_CHECK_EQUAL(p.inCorrectSubgroup(), false);
+    g2 p_res = g2::fromJacobianBytesBE(hexToBytes<288>("0x158a2a1e3ce68c49f9795908aa3779c6919ed5de5cbcd1d2a331d0742d1eb3cb28014006b5f686204adb5fdca73aea570ee0f0d58880907c8de5867dd99b6b7306b2c3de4a1537e6d042f2b8e44c8086853728cc246726016b0fcf993db3d759005f8ac0cb55113c857c5cf3f83d9b624ce9a2a0a00a1206777cf935721c857b322a611ed0703cf3e922bfb8b19a1f5e10a341b2191ab5a15d35f69850d2adb633e5425eecb7f38dd486a95b3f74d60f3ee6cf692b3c76813407710630763f7605b3828c19203f661732a02f7f546ab354694128bbe5a792a9db4a443c0fe10af0df2bc1b8d07aee99bd6f8c6b26847011aa31634f42f722d52022c736369db470576687fdf819cf15a0db4c01a0bd7028ee17cefdf6d66557d47fb725b6d00f"), false, true).value();
+    BOOST_CHECK_EQUAL(p.equal(p_res), true);
+} FC_LOG_AND_RETHROW();
```

Comment:
These tests of garbage input are good, but I would like to see unit tests that directly test the code paths of the host functions for: these garbage cases; failure cases that return -1; and even one successful case for each function (taken from the libraries test vectors).

This will require adding a new test contract.

See how it was done for the alt_bn_128 host functions: https://github.com/eosnetworkfoundation/mandel/pull/316/files#diff-87c8848613f6d5f711354edd61feeebcaae9b6acd4f5f16ca651075b989dc929

Reply:
yeah, that failure is because the host function is named `bls_pairing` which doesn't match,
https://github.com/AntelopeIO/leap/blob/712b1a0faa2438dcc28b77bd807c0f1b99b02dd8/libraries/chain/include/eosio/chain/webassembly/eos-vm-oc/intrinsic_mapping.hpp#L277
Fixing the entry in intrinsic_mapping.hpp will resolve it.

(in 2.0 this mismatch would have ended up as a compile error, we lost that somewhere between 2.0 & 3.1 -- #621)

Reply:
Thanks, Matt. Fixed in: https://github.com/AntelopeIO/leap/pull/1071/commits/59cc737a42611ede24d423aaf5fff6c578de039e